sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -243,17 +247,36 @@ class FindDataSourceTable(sparkSession: SparkSession) extends Rule[LogicalPlan]
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
     case i @ InsertIntoStatement(UnresolvedCatalogRelation(tableMeta), _, _, _, _)
         if DDLUtils.isDatasourceTable(tableMeta) =>
-      i.copy(table = readDataSourceTable(tableMeta))
+      if (DataSource.isV2Provider(tableMeta.provider.get, sparkSession.sessionState.conf)) {
```

Comment:
I see the problem now. The table lookup for SELECT/INSERT is more complicated than I thought:
1. try to lookup temp view first.
2. lookup table/view. If it's a table from the session catalog, we should create a v1 relation if table provider is v1, otherwise create v2 relation.

In fact, we rely on the order of `ResolveTables` and `ResolveRelations`, which is pretty bad and violates the design of catalyst. The rules in one batch should be order-insensitive.

This fix does resolve the problem: even if we mistakenly resolve to a v1 relation, we still have a chance to correct it to a v2 relation. But I think it's better to fix the root cause: `ResolveTables` and `ResolveRelations` are order-sensitive.

Reply:
I've proposed https://github.com/apache/spark/pull/26214 to fix this problem.

Reply:
Would you mind getting this one in first and then discussing how to fix the order problem? I'd like to get early pushdown fixed and then handle the issues that we've found.

Reply:
hmm I'm a little confused. We hit the problem because we refactor the table lookup rules in this PR. IIUC this PR only needs to add resolving logic of `DescribeTable` and `AlterTable` to `ResolveTables`, and we can get rid of this problem if we don't refactor the table lookup rules.

Reply:
My point is that this is currently working and I'd rather not wait until we find a different solution to the rule ordering problem to get this work in.

The rules are dependent on order whether or not the refactor is done.

Reply:
Yes it's currently working. But it adds a refactor and a hack to fix the issue introduced by the refactoring.

We should not wait until we find a different solution, but it's better to not expose the problem in this PR.

That said, the refactor in this PR may be problematic. More specifically, converting `UnresolvedRelation` to `UnresolvedV2Relation` in `ResolveCatalogs` may not be a good idea. I think we should either 1) revert the refactor in this PR, or 2) wait for #26214 until we reach an agreement about how `UnresolvedRelation` should be resolved.

Reply:
> No, it fixes an existing problem -- that rules were order dependent.

Yes, this is an existing problem, but does it affect this patch? IIUC this PR only needs to add resolving logic of `DescribeTable` and `AlterTable` to `ResolveTables`, so it doesn't matter that rules are order dependent.

The refactor that is done in this PR, exposes the problem. I think we should revert that refactor, and only keep the "early pushdown change".

More specifically, let's revert the changes made in `ResolveCatalogs` and `ResolveSessionCatalog`. We should keep `ResolveTables` to still resolve both `UnresolvedRelation` and `UnresolvedV2Relation`. Please treat this as a -1.

Reply:
Fair enough, I'll remove the refactor.